### DISCUSSION TITLE
Photoshop 8bf plugin support

### DISCUSSION THREAD

**Discussion by tannerhelland** [tannerhelland/PhotoDemon#351] [News]
As of February 2021, PhotoDemon nightly builds now attempt to support Photoshop 8bf (filter) plugins:

![image](https://user-images.githubusercontent.com/1930029/108232022-9729c300-70ff-11eb-8ba3-78895ab5f7f8.png)

This feature is possible thanks to the very cool 3rd-party "Photoshop Plugin Host" library by @spetric:

https://github.com/spetric/Photoshop-Plugin-Host

Photoshop plugins are a new area of exploration for me, and most my testing so far has been against 8bf files that ship with old copies of Photoshop.  (These plugins appear to work very well, yay!)

At this point, I'm far more worried about non-Adobe plugins, because they're much more likely to contain bugs or use nonstandard behavior.  If any PhotoDemon users have a collection of 3rd-party 8bf plugins (or are willing to download some for testing), I'd love to get feedback on this new feature.  

If you encounter any plugins that don't work, please file an issue or report the issues in this thread.  The name of the plugin involved (and where I can find it online, if relevant) would also help a ton.

Thank you!

**Comment by ntstatic** [comment]
Thank you ... and since we are discussing it thanks for all the work put in to this excellent piece of art 

Where can i download the Nightly Build From

**Reply by ntstatic** [reply]
ok i got it ... thanks once again

**Comment by fvbeers** [comment]
Great addition,  I use Topaz Denoise AI al lot, now in PhotoDemon too! :-)

**Comment by user-by** [comment]
How to work gmic ps plugins? I tried 32bit version but didn't work.

**Comment by spetric** [comment]
I haven't tried GMIC ps filter, but as I recall GMIC uses 32-bit floating point channels and current plug-in host version does not support floating point channels.I have some updates (bug corrections) to upload on Github and then I can build new DLLs and try how it works with GMIC.    
    On Saturday, August 5, 2023 at 01:23:57 PM GMT+2, user-by ***@***.***> wrote:  
 
 


How to work gmic ps plugins? I tried 32bit version but didn't work.

**Reply by user-by** [reply]
Ok thanks, i will wait, also can you check these plug-ins support. Because getting memory error. 
https://www.vicanek.de/imageprocessing/contrastmask.htm

**Reply by 0xC0000054** [reply]
The problem with the Contrast Mask filter is that PhotoDemon enables DEP, and many 3rd party Photoshop plugins are not compatible with it. For Contrast Mask, the FilterMeister [FMPatcher](https://www.filtermeister.com/fmpatcher/index.htm) tool should fix the DEP issue.
The underlying issue is that those plugins are executing code in memory that is not marked as executable, which is what DEP is designed to prevent.

As for the GMIC plugin, it is a bug in how the host library scans for plugins. The GMIC plugin contains two plugins in the same file, and the host is only returning the last plugin it finds. I opened an issue for this bug at: https://github.com/spetric/Photoshop-Plugin-Host/issues/3

I am the author of the [GMIC 8bf filter](https://github.com/0xC0000054/gmic-8bf) and the [Paint.NET Photoshop plugin host](https://github.com/0xC0000054/PSFilterPdn).

**Reply by tannerhelland** [reply]
Thank you so much for this helpful info, @0xC0000054 .  I apologize that I didn't see your comment until now.

Years ago, I manually enabled DEP for PhotoDemon in hopes of reducing false-positives with various virus scanners, but it's (obviously) not required for the app to work correctly.  I will try turning off DEP and seeing if it makes a material difference in false-positive reports.  The benefit of increased PS plugin compatibility seems worth it.

Thank you again for your kind help.  (And thank you for all your work on Paint.NET plugins - I use and appreciate many of them!)

**Reply by 0xC0000054** [reply]
No problem. There currently has been no response to the bug report I filed about 8bf filters that have multiple plugins in the same file.
I looked at the `pspiHost.dll` filter search code, but I could not determine the best way to modify it to support multiple filters in the same file.

As a side note, my Paint.NET 8bf filter plugin implements almost all of the features that `pspiHost.dll` lists as currently not supported (padding, parameters, etc). 
The only thing I don't implement is the `Big document support`, which was introduced in Photoshop 8.0 (CS).
Adobe changed the Photoshop SDK license after version 6.0 to prevent 3rd-party host developers from using the new APIs.
